template.tpl

```diff
@@ -1735,69 +1739,99 @@ const mapSnowplowTp2EventToTagEvent = (event, config) => {
     withCurrentEnv(addSelfDescPropsTp2),
     withCurrentEnv(populateAdditionalProperties),
     withCurrentEnv(postprocess),
-    cleanObject
-  ].reduce((acc,curr) => curr(acc), commonEvent);
+    cleanObject,
+  ].reduce((acc, curr) => curr(acc), commonEvent);
 
   return result;
 };
 
-const requestParts = REQUEST_PATH.split('/');
+/**
+ * Returns the CDN URL of the requested JS tracker version.
+ *
+ * @param {string} - The requested JS tracker version
+ * @returns {string} The URL
+ */
+const getSpJsLocation = (spJsVersion) => {
+  const file = '/sp.js';
+  if (makeInteger(spJsVersion.charAt(0)) <= 2) {
+    return CDN.concat('/gh/snowplow/sp-js-assets@', spJsVersion, file);
+  }
 
-if (requestParts.length > 2) {
-  const requestedSpJsVersion = requestParts[1];
-  const requestedSpJsName = requestParts[2];
+  return CDN.concat(
+    '/npm/@snowplow/javascript-tracker@',
+    spJsVersion,
+    '/dist',
+    file
+  );
+};
 
-  if (
-    data.serveSpJs &&
-    (requestedSpJsName === data.customSpJsName || requestedSpJsName === 'sp.js')
-  ) {
-    claimRequest();
-    log('Snowplow sp.js request, claimed...');
+/**
+ * Handles the request for JS tracker library.
+ *
+ * @param {Array} requestPathParts - The request path's parts
+ * @param {Object} cfg - The client configuration
+ */
+const handleSpJsRequest = (requestPathParts, cfg) => {
+  const requestedSpJsVersion = requestPathParts[1];
+  const requestedSpJsName = requestPathParts[2];
+  const cacheVersionKey = 'snowplow_gtm_ss_client_version';
+  const contentKey = 'snowplow_js_' + requestedSpJsVersion;
+  const headersKey = 'snowplow_js_headers_' + requestedSpJsVersion;
+
+  // clear cache if necessary
+  const cacheVersion = templateDataStorage.getItemCopy(cacheVersionKey);
+  if (cacheVersion !== VERSION) {
+    templateDataStorage.clear();
+    templateDataStorage.setItemCopy(cacheVersionKey, VERSION);
+  }
 
-    const cachedSpJs = templateDataStorage.getItemCopy(
-      'snowplow_js_' + requestedSpJsVersion
+  const cachedSpJs = templateDataStorage.getItemCopy(contentKey);
+  const cachedSpJsHeaders = templateDataStorage.getItemCopy(headersKey) || {};
+  if (!cachedSpJs) {
+    const spJsLocation = getSpJsLocation(requestedSpJsVersion);
+    sendHttpGet(
+      spJsLocation,
+      (statusCode, headers, body) => {
+        if (statusCode >= 200 && statusCode < 300) {
+          templateDataStorage.setItemCopy(contentKey, body);
+          templateDataStorage.setItemCopy(headersKey, headers);
+          sendResponse(200, body, headers);
+        } else {
+          log('Failed to download sp.js: ', body);
+          sendResponse(statusCode, body, headers);
+        }
+      },
+      { timeout: 5000 }
```

Comment:
It complicates the implementation quite a bit but at some point we should consider handling (separately caching) `accept-encoding` to send the actual compressed file for perf reasons.